File: yt/yt/flow/examples/python/async_request/state_keeper_function.py
"""StateKeeperFunction: handles events and responses."""

import logging
import random

from yt.yt.flow.library.python.companion.computation import RowFunction

log = logging.getLogger(__name__)


# [BEGIN state_keeper]
class StateKeeperFunction(RowFunction):
    """Sends requests for events; accumulates total_length for responses."""

    def on_message(self, message, output, ctx):
        stream_id = message.stream_id

        if stream_id == "event":
            key = message.payload["key"]
            data = message.payload["data"]
            request_id = random.getrandbits(64)

            builder = ctx.message_builder("request")
            builder.set("request_id", request_id)
            builder.set("key", key)
            builder.set("request", data)
            output.add_message(builder.finish())
            log.debug("Send request (RequestId: %s)", request_id)

        elif stream_id == "response":
            length = message.payload["length"]
            request_id = message.payload["request_id"]
            log.debug("Received response (RequestId: %s)", request_id)

            state = ctx.external_state("/state", message)
            current = state.get("total_length")
            total = (current if current is not None else 0) + length
            builder = state.to_builder()
            builder.set("total_length", total)
            state.set(builder.finish())
        else:
            raise ValueError(f"Unknown stream_id: {stream_id}")


# [END state_keeper]

File: yt/yt/flow/examples/kotlin/retryable_async_request/retryable_async_request/src/main/kotlin/tech/ytsaurus/flow/examples/retryableasyncrequest/StateKeeperFunction.kt
package tech.ytsaurus.flow.examples.retryableasyncrequest

import org.slf4j.Logger
import org.slf4j.LoggerFactory
import tech.ytsaurus.flow.computation.OutputCollector
import tech.ytsaurus.flow.context.RuntimeContext
import tech.ytsaurus.flow.function.RowFunction
import tech.ytsaurus.flow.row.ExtendedMessage
import tech.ytsaurus.flow.row.Payload
import tech.ytsaurus.flow.spring.FlowComputation
import tech.ytsaurus.flow.state.ExternalStateAccessor
import tech.ytsaurus.flow.state.ExternalStateDescriptor
import tech.ytsaurus.flow.state.StateDescriptors
import java.util.Random

// [BEGIN registration]
@FlowComputation(id = "state")
// [END registration]

/**
 * Handles both "event" and "response" streams:
 *   - event  -> creates a request message and emits it to the "request" stream
 *   - response -> accumulates total_length in external state ("state" table)
 *
 * Identical logic to the async_request StateKeeper.
 */
open class StateKeeperFunction : RowFunction {
    companion object {
        private val log: Logger = LoggerFactory.getLogger(StateKeeperFunction::class.java)

        private val RANDOM = Random()

        private val STATE: ExternalStateDescriptor =
            StateDescriptors.external("/state")
    }

    // [BEGIN on_message]
    override fun onMessage(message: ExtendedMessage, output: OutputCollector, ctx: RuntimeContext) {
        val streamId = message.streamId

        when (streamId) {
            "event" -> handleEvent(message, output, ctx)
            "response" -> handleResponse(message, ctx)
            else -> throw IllegalArgumentException("Unknown streamId: $streamId")
        }
    }
    // [END on_message]

    private fun handleEvent(message: ExtendedMessage, output: OutputCollector, ctx: RuntimeContext) {
        val key = message.get("key", Long::class.java)
        val data = message.get("data", String::class.java)
        val requestId = RANDOM.nextLong().ushr(1)

        log.debug("Handling event (key={}, requestId={})", key, requestId)

        val requestBuilder = ctx.createMessageBuilder("request")
        requestBuilder.set("request_id", requestId)
        requestBuilder.set("key", key)
        requestBuilder.set("request", data)

        output.addMessage(requestBuilder.finish())
    }

    private fun handleResponse(message: ExtendedMessage, ctx: RuntimeContext) {
        val length = message.get("length", Long::class.java)!!
        val key = message.get("key", Long::class.java)

        log.debug("Handling response (key={}, length={})", key, length)

        val stateAccessor: ExternalStateAccessor = ctx.getState(STATE, message)
        val current: Payload = stateAccessor.getOrDefault()
        val updated = current.toBuilder()

        val totalLength = current.get("total_length", Long::class.java) ?: 0L
        updated.set("total_length", totalLength + length)

        stateAccessor.set(updated.finish())
    }
}
